kong/tools/queue.lua

```diff
@@ -0,0 +1,369 @@
+--
+-- This is a queue of "entries". Entries can be any Lua value.
+--
+-- The queue has two mandatory parameters:
+--
+-- `name` is the name of the queue. It uniquely identifies the queue across the
+-- system.  If two plugin instances use the same queue name, they will share one
+-- queue and their queue related configuration must match.
+--
+-- `handler` is the function which consumes entries.
+--
+-- Entries are dequeued in batches.  The maximum size of each batch of entries
+-- that is passed to the `handler` function can be configured using the
+-- `batch_max_size` parameter.
+--
+-- The handler function can either return true if the entries
+-- were successfully processed, or false and an error message to indicate that
+-- processing has failed.  If processing has failed, the queue library will
+-- automatically retry.
+--
+-- If the `batch_max_size` parameter is larger than 1, processing of the
+-- queue will only start after `max_delay` milliseconds have elapsed if
+-- less than `batch_max_size` entries are waiting on the queue.  If `batch_max_size`
+-- is 1, the queue will be processed immediately.
+--
+-- Usage:
+--
+--   local Queue = require "kong.tools.queue"
+--
+--   local handler = function(entries)
+--     -- must return true if ok, or false + error otherwise
+--     return true
+--   end
+--
+--   local q = Queue.new(
+--     "example",
+--     handler, -- function used to process values from the queue
+--     { -- Opts table with control values. Defaults shown:
+--       retry_count      = 0,    -- number of times to retry processing
+--       batch_max_size   = 1000, -- max number of entries that handed to the process function in one invocation
+--       max_delay        = 1000, -- processing delay in milliseconds
+--     }
+--   )
+--
+--   q:add("Some value")
+--   q:add("Some other value")
+--
+--   ...
+--
+-- Given the example above,
+--
+-- * If the two `q:add()` invocations are done in quick succession, they will be passed to the
+--   handler function together in one batch.
+-- * If processing fails, it will not be retried (retry_count equals 0)
+-- * If retry_count was bigger than 0, processing would be re-queued n times before finally being discarded.
+-- * The retries are not regular: every time the handler fails, they next retry is delayed by n_try^2, up to
+--   max_retry_delay seconds.
+--
+-- The most important internal attributes of Queue are:
+--
+-- * `self.queue`: This is an array of entries, which are awaiting processing/consumption
+--
+-- Each entry has the following structure:
+-- {
+--   data = "some data",     -- data to process
+--   config = {...},         -- plugin configuration that caused the entry to be created
+--   timestamp = 12345789,   -- timestamp when the entry was queued
+-- }
+
+
+local semaphore = require "ngx.semaphore"
+
+local DEBUG = ngx.DEBUG
+local INFO = ngx.INFO
+local WARN = ngx.WARN
+local ERR = ngx.ERR
+
+
+local function now()
+  ngx.update_time()
+  return ngx.now()
```

Comment:
localzie `ngx.update_time` `ngx.now`

Reply:
Localizing won't have a noticeable effect on performance, but would make the code harder to understand and maintain.

Reply:
@hanshuebner Can't fully agree with this statement. First, not localizing deviates from established styles of the repo and best practices. Second, small differences do adds up.